### PR TITLE
MM-14417: Add messaging for channel mentions of users not in constrai…

### PR DESCRIPTION
**Pull request by mkraft**
…ned groups.

#### Summary

Updates the display of the ephemeral message notifying of users mentioned in a channel when the channel is group-constrained and some or all of the mentioned users are not permitted in the channel.

Requires this server change: https://github.com/mattermost/mattermost-server/pull/10594

The equivalent web app PR is: https://github.com/mattermost/mattermost-webapp/pull/2647

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-14417

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [ ] All new/modified APIs include changes to [mattermost-redux](https://github.com/mattermost/mattermost-redux) (please link)
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information

This PR was tested on: iPhone X, iOS 12.2 (emulated)

#### Screenshots

Simulator Screen Shot - iPhone X - 2019-04-11 at 16 49 28
Simulator Screen Shot - iPhone X - 2019-04-11 at 16 49 34

### CODE REVIEW

File: app/components/post_add_channel_member/post_add_channel_member.js [hunk 1]
```diff
@@ -150,43 +152,75 @@ export default class PostAddChannelMember extends React.PureComponent {
...
+                        id={outOfGroupsMessageID}
+                        defaultMessage={outOfGroupsMessageText}
+                        style={baseTextStyle}
+                    />
+                </Text>
+            );
+        }
+
+        return (
+            <>
```
Comment (enahum) [inline app/components/post_add_channel_member/post_add_channel_member.js]: what is this empty `<>`  tag? I think is safer to use `<React.Fragment>` instead.

**Comment by enahum** [comment]
@mkraft can you post an example where the groups options is not enabled?

File: app/components/post_add_channel_member/post_add_channel_member.js (hunk 1 above)
Comment (mkraft) [reply app/components/post_add_channel_member/post_add_channel_member.js]: @enahum `<>` is the short form of `<React.Fragment>`.

Comment (enahum) [reply app/components/post_add_channel_member/post_add_channel_member.js]: you always learn something new ;) thanks my friend.

In any case please post a screenshot to see if possible without the groups stuff enabled.

**Comment by mkraft** [comment]
@enahum Here's are two notifications, one with a single user and the other with multiple users.

Simulator Screen Shot - iPhone X - 2019-04-15 at 11 44 01

**Comment by enahum** [comment]
@mkraft I mean for a server that does not have the option or that the option is NOT enabled, or even a channel that is not controlled by any groups.

**Comment by mkraft** [comment]
@enahum The screenshot is for a group that is not group-constrained. Am I misunderstanding?

**Comment by enahum** [comment]
@mkraft so basically what I want to see if how the app behaves with backwards compatibility

**Comment by mkraft** [comment]
@enahum Here's the master branch of mattermost-mobile pointing to the server-side changes associated to MM-14417 using a non-group-constrained private channel:

Simulator Screen Shot - iPhone X - 2019-04-15 at 12 14 43

**Comment by enahum** [comment]
So this means that this PR pointing to a server that has the  MM-14417 changes will always show that they do not belong to a group even though LDAP groups is not enabled?

**Comment by mkraft** [comment]
> So this means that this PR pointing to a server that has the MM-14417 changes will always show that they do not belong to a group even though LDAP groups is not enabled?

I don't understand what you mean. Where does it show that they don't belong to a group? Do you mean belong to a channel? The users in that example don't belong to the channel.

File: app/components/post_add_channel_member/post_add_channel_member.js [hunk 2]
```diff
@@ -30,6 +30,7 @@ export default class PostAddChannelMember extends React.PureComponent {
         postId: PropTypes.string.isRequired,
         userIds: PropTypes.array.isRequired,
         usernames: PropTypes.array.isRequired,
+        noGroupsUsernames: PropTypes.array.isRequired,
```
Comment (mkraft) [inline app/components/post_add_channel_member/post_add_channel_member.js]: This cannot be required for backwards compatibility reasons.

File: app/components/post_add_channel_member/post_add_channel_member.js [hunk 3]
```diff
@@ -150,43 +152,75 @@ export default class PostAddChannelMember extends React.PureComponent {
...
-                    id={'post_body.check_for_out_of_channel_mentions.message_last'}
-                    defaultMessage={'? They will have access to all message history.'}
-                    style={baseTextStyle}
-                />
-            </Text>
+            );
+        }
+
+        let outOfGroupsMessage = null;
+        if (noGroupsUsernames.length) {
```
Comment (mkraft) [inline app/components/post_add_channel_member/post_add_channel_member.js]:
```suggestion
        if (noGroupsUsernames?.length) {
```

**Comment by mkraft** [comment]
@enahum Thanks for reminding me to also test an upgraded mobile app with an old server. Here's a screenshot of it working:

Simulator Screen Shot - iPhone X - 2019-04-15 at 13 56 02

Now I have covered the following:

- [x] Upgraded mobile app, upgraded server 
- [x] Existing mobile app, upgraded server
- [x] Upgraded mobile app, existing server

File: app/components/post_add_channel_member/post_add_channel_member.js (hunk 1 above)
Comment (mgdelacroix) [reply app/components/post_add_channel_member/post_add_channel_member.js]: Today I Learned... :D

**Pull request merged**